feat: add form utility components for easier form handling

sidarth-23 opened the pull request:
<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. 
-->

## Summary
- The react-hook-form library was added as a dependency to enable form validation and state management.
- Added `FormItem` component to provide a context for form items
- Added `FormMessage` component to display error messages for form fields
- Added `ReactForm` object to export all form-related components
- Added `FormLabel` component to handle form labels with error styling
- Added `Form` component as a form provider
- Added `FormField` component to handle form fields with context
- Added `useFormField` hook to access form field context and state

### UPDATED ANSWERS TO COMMENTS
- The TFieldName is the best use of naming as the use of the type is a generic
- Similarly to other places where T is used

### FUTURE PLAN
- To keep this component untouched for future use in other forms, and add a abstraction for the auth pages for design consistency

<!--
 Explain the **motivation** of this Pull Request
-->
### Checklist
- [x] Tested (Must)
- [ ] Test Case added
- [x] Build Successful (Must)
- [x] Sufficient Code comments added (Must)
- [ ] Attached Screenshots / Videos <!-- if the PR contains UI changed, fixes, improvements -->
- [x] All Relevant Documents added


## Depends on
#41 
<!--- Does this PR depend on another PR that should be merged first or at the same time -->

yaju09 commented on packages/ui/src/partials/react-forms.tsx:
```diff
@@ -0,0 +1,176 @@
+const Form = FormProvider
+
+type FormFieldContextValue<
+  TFieldValues extends FieldValues = FieldValues,
```
```suggestion
  TFieldValues extends FieldValues = FieldValues,
```
can we come up with a more meaningful name for the same?

yaju09 commented on packages/ui/src/partials/react-forms.tsx:
```diff
@@ -0,0 +1,176 @@
+  TFieldValues extends FieldValues = FieldValues,
+  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
+> = {
+  name: TName
```
```suggestion
  name: TName
```

can we come up with a better name?

yaju09 commented on packages/ui/src/partials/react-forms.tsx:
```diff
@@ -0,0 +1,176 @@
+  {} as FormItemContextValue
+)
+
+const FormItem = React.forwardRef<
```
```suggestion
const FormItem = React.forwardRef<
```
can we create separate comps for different form items like formItem, formlabel, form description etc?

yaju09 commented on packages/ui/src/partials/index.ts:
```diff
@@ -3,3 +3,4 @@ export * from './card'
 export * from './checkbox'
 export * from './input'
 export * from './label'
+export * from './react-form/react-form'
```
```suggestion
export * from './react-form/form'
```

will it be better to rename the form comp as form?

sidarth-23 replied:
This can be done

yaju09 commented on packages/ui/src/partials/react-form/form-field.tsx:
```diff
@@ -0,0 +1,81 @@
+ * This is an extensible type that can be used to add additional properties to the form field context
+ * */ 
+type FormFieldContextValue<
+  TFieldValues extends FieldValues = FieldValues,
```
```suggestion
  TFieldValues extends FieldValues = FieldValues,
```
can we have a more meaningful name for the TFieldValues

sidarth-23 replied:
This is a generic. Hence I think it is norm that generics start with a T in the beginning

https://wanago.io/2020/02/17/typescript-generics-discussing-naming-conventions/

yaju09 commented on packages/ui/src/partials/react-form/form-field.tsx:
```diff
@@ -0,0 +1,81 @@
+ * */ 
+type FormFieldContextValue<
+  TFieldValues extends FieldValues = FieldValues,
+  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
```
```suggestion
  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
```
more meaningful name for TName

sidarth-23 replied:
This is a generic. Hence I think it is norm that generics start with a T in the beginning

https://wanago.io/2020/02/17/typescript-generics-discussing-naming-conventions/

yaju09 commented on packages/ui/src/partials/react-form/form-item.tsx:
```diff
@@ -0,0 +1,24 @@
+  id: string
+}
+
+export const FormItemContext = React.createContext<FormItemContextValue>(
```
can the context be defined in one central react function and imported from there wherever required?

sidarth-23 replied:
This is not going to exported directly to be consumed. So is it worth it creating a separate file for this

yaju09 commented on packages/ui/src/partials/react-form/form-field.tsx:
```diff
@@ -0,0 +1,81 @@
+ *
+ * @default {} as FormFieldContextValue
+ */
+const FormFieldContext = React.createContext<FormFieldContextValue>(
```
can the context be defined in one central react function and imported from there wherever required?

sidarth-23 replied:
This is not going to exported directly to be consumed. So is it worth it creating a separate file for this

yaju09 commented on packages/ui/src/partials/react-form/form-field.tsx:
```diff
@@ -0,0 +1,81 @@
+ * 
+ * @returns The form field context
+ */
+export const useFormField = () => {
```
can we create a separate custom in a diff comp?

sidarth-23 replied:
I think I din't want to to abstract out too much. So as this is related to the field, I kept it in the same file

yaju09 reviewed (changes requested):
how will  it be to define all the contexts centrally within the react form directory and import then wherever needed ?
Also, how about defining custom hooks separate comps?

sidarth-23 committed:
feat: add form utility components for easier form handling

- Added a set of form utility components including Form, FormItem, FormLabel, FormControl, FormDescription, FormMessage, and FormField.
- These components provide a convenient way to handle form fields, labels, descriptions, and error messages.
- The react-hook-form library was added as a dependency to enable form validation and state management.
feat: add new components for form handling and form validation

- Added `FormItem` component to provide a context for form items
- Added `FormMessage` component to display error messages for form fields
- Added `ReactForm` object to export all form-related components
- Added `FormLabel` component to handle form labels with error styling
- Added `Form` component as a form provider
- Added `FormField` component to handle form fields with context
- Added `useFormField` hook to access form field context and state
feat: update dependencies and reorganize form components

- Updated dependencies to include "@hookform/resolvers" version 3.3.4
- Reorganized form components into separate files for better organization and maintainability

added previous commit to prevent merge conflict
